File: src/services/housekeepingService.js
/**
 * housekeepingService.js
 * --------------------------------------------------------------------------
 * Tozalash Servisi. Vazifalar:
 *   - 'guest.checked_out' va 'room.cleaning_required' hodisalarini qabul qilish
 *   - Xonalarni tozalash navbatiga (FIFO) qo'shish
 *   - Tozalovchilarga avtomatik tayinlash
 *   - Holat o'tishlari: cleaning_required -> cleaning -> inspection
 *
 * Eslatma: markClean endi xonani 'available' emas, 'inspection' holatiga
 * o'tkazadi. Qabul xodimi tasdiqlaganidan keyin 'available' bo'ladi.
 * --------------------------------------------------------------------------
 */

'use strict';

const broker = require('../broker/messageBroker');
const store = require('../data/store');
const logger = require('../utils/logger');

class HousekeepingService {
  constructor() {
    this.name = 'housekeeping';
    this.cleaningQueue = []; // FIFO navbat
    this._subscribe();
    this._restoreQueueFromStore();
    logger.info('[HOUSEKEEPING] Servis ishga tushdi');
  }

  _subscribe() {
    // Mehmon check-out qilganda — checkOut allaqachon statusni o'zgartirgan,
    // biz faqat navbatga qo'shamiz
    broker.subscribe('guest.checked_out', (event) => {
      const roomNumber = event.payload.room?.number;
      if (roomNumber) {
        this._enqueueWithoutStatusChange(roomNumber, 'guest_checkout');
      }
    });

    // 12-soatlik tekshiruvdan
    broker.subscribe('room.cleaning_required', (event) => {
      const roomNumber = event.payload.roomNumber;
      this._enqueueWithoutStatusChange(roomNumber, 'periodic_check');
    });
  }

  /** Server qayta ishga tushganda mavjud cleaning_required xonalarni navbatga qo'shamiz */
  _restoreQueueFromStore() {
    for (const r of store.getRooms()) {
      if (r.status === 'cleaning_required') {
        this.cleaningQueue.push({
          roomNumber: r.number,
          reason: 'restored',
          addedAt: r.dirtyAt || Date.now(),
        });
      }
    }
  }

  /** Faqat navbatga qo'shadi, statusni o'zgartirmaydi (status allaqachon tozalash_kk) */
  _enqueueWithoutStatusChange(roomNumber, reason) {
    const room = store.getRoom(roomNumber);
    if (!room) return null;
    if (room.status !== 'cleaning_required') return null;
    if (this.cleaningQueue.find((q) => q.roomNumber === roomNumber)) return null;

    const entry = { roomNumber, reason, addedAt: Date.now() };
    this.cleaningQueue.push(entry);
    this._tryAssignCleaner(entry);
    return entry;
  }

  /**
   * Qo'lda navbatga qo'shish (Reception yoki manager tomonidan).
   * Statusni cleaning_required qiladi (agar boshqacha bo'lsa).
   */
  addToCleaningQueue(roomNumber, reason = 'manual') {
    const room = store.getRoom(roomNumber);
    if (!room) return null;
    if (room.status === 'occupied') {
      logger.warn(`[HOUSEKEEPING] ${roomNumber}-xona band, navbatga qo'shilmaydi`);
      return null;
    }
    if (room.status === 'cleaning' || room.status === 'maintenance') {
      return null;
    }
    if (this.cleaningQueue.find((q) => q.roomNumber === roomNumber)) {
      return null;
    }

    // Status 'available' yoki 'inspection' bo'lsa va sabab 'guest_checkout' bo'lsa, bekor qilamiz
    if (room.status !== 'cleaning_required' && reason === 'guest_checkout') {
      logger.debug(`[HOUSEKEEPING] ${roomNumber} allaqachon ${room.status}, navbatga qo'shilmaydi`);
      return null;
    }

    // Status cleaning_required emas bo'lsa, qilamiz
    if (room.status !== 'cleaning_required') {
      const oldStatus = room.status;
      store.updateRoom(roomNumber, { status: 'cleaning_required', dirtyAt: Date.now() });
      broker.publish('room.status_changed', {
        roomNumber, oldStatus, newStatus: 'cleaning_required',
        changedBy: 'housekeeping', reason,
      });
    }

    const entry = { roomNumber, reason, addedAt: Date.now() };
    this.cleaningQueue.push(entry);
    this._tryAssignCleaner(entry);
    return entry;
  }

  _tryAssignCleaner(queueEntry) {
    const housekeepers = store.getHousekeepers();
    const free = housekeepers.find((h) => h.available);
    if (!free) return null;
    queueEntry.assignedTo = free.id;
    queueEntry.assignedToName = free.name;
    queueEntry.assignedAt = Date.now();
    return free;
  }

  /** Tozalashni boshlash: cleaning_required -> cleaning */
  startCleaning(roomNumber, byUser = 'system') {
    const room = store.getRoom(roomNumber);
    if (!room) return { success: false, error: `${roomNumber}-xona topilmadi` };
    if (room.status !== 'cleaning_required') {
      return { success: false, error: `${roomNumber}-xona tozalash uchun tayyor emas (joriy: ${room.status})` };
    }
    const oldStatus = room.status;
    store.updateRoom(roomNumber, {
      status: 'cleaning',
      cleaningStartedAt: Date.now(),
      cleanedBy: byUser,
    });
    broker.publish('room.status_changed', {
      roomNumber, oldStatus, newStatus: 'cleaning',
      changedBy: 'housekeeping', actor: byUser,
    });
    logger.info(`[HOUSEKEEPING] ${roomNumber}-xona tozalanmoqda (${byUser})`);
    return { success: true, room: store.getRoom(roomNumber) };
  }

  /** Tozalashni yakunlash: cleaning -> inspection (qabul tekshirishi kerak) */
  markClean(roomNumber, byUser = 'system') {
    const room = store.getRoom(roomNumber);
    if (!room) return { success: false, error: `${roomNumber}-xona topilmadi` };
    if (room.status !== 'cleaning' && room.status !== 'cleaning_required') {
      return { success: false, error: `${roomNumber}-xona tozalanmagan (joriy: ${room.status})` };
    }
    const oldStatus = room.status;
    const cleaningDurationMs = room.cleaningStartedAt ? Date.now() - room.cleaningStartedAt : null;

    store.updateRoom(roomNumber, {
      status: 'inspection',
      lastCleanedAt: Date.now(),
      inspectionStartedAt: Date.now(),
      dirtyAt: null,
      cleaningStartedAt: null,
      lastCleaningDurationMs: cleaningDurationMs,
      lastCleanedBy: byUser,
    });

    // Navbatdan olib tashlaymiz
    this.cleaningQueue = this.cleaningQueue.filter((q) => q.roomNumber !== roomNumber);

    // Statistikani yangilaymiz
    store.incrementStat('totalCleanings');

    broker.publish('room.status_changed', {
      roomNumber, oldStatus, newStatus: 'inspection',
      changedBy: 'housekeeping', actor: byUser,
    });
    broker.publish('notification.created', {
      type: 'inspection_required',
      severity: 'info',
      message: `${roomNumber}-xona tozalandi va qabul tomonidan tekshirilishi kutilmoqda`,
      roomNumber,
    });

    logger.info(`[HOUSEKEEPING] ${roomNumber}-xona tozalandi -> TEKSHIRUVDA`);
    return { success: true, room: store.getRoom(roomNumber) };
  }

  getQueue() {
    return this.cleaningQueue.map((q) => ({
      ...q,
      room: store.getRoom(q.roomNumber),
    }));
  }
}

module.exports = new HousekeepingService();

File: backend/app/services/housekeeping_service.py
import time
from app.broker.message_broker import broker
from app.data.store import store
from app.utils.logger import logger

class HousekeepingService:
    def __init__(self):
        self.name = 'housekeeping'
        self.cleaning_queue = []
        self._subscribe()
        self._restore_queue_from_store()
        logger.info('[HOUSEKEEPING] Servis ishga tushdi')

    def _subscribe(self):
        def on_guest_checked_out(event):
            room = event.get('payload', {}).get('room', {})
            room_number = room.get('number')
            if room_number is not None:
                self._enqueue_without_status_change(room_number, 'guest_checkout')

        def on_room_cleaning_required(event):
            room_number = event.get('payload', {}).get('roomNumber')
            if room_number is not None:
                self._enqueue_without_status_change(room_number, 'periodic_check')

        broker.subscribe('guest.checked_out', on_guest_checked_out)
        broker.subscribe('room.cleaning_required', on_room_cleaning_required)

    def _restore_queue_from_store(self):
        for r in store.get_rooms():
            if r.get('status') == 'cleaning_required':
                self.cleaning_queue.append({
                    "roomNumber": r['number'],
                    "reason": 'restored',
                    "addedAt": r.get('dirtyAt', int(time.time() * 1000)),
                })

    def _enqueue_without_status_change(self, room_number, reason):
        room = store.get_room(room_number)
        if not room or room.get('status') != 'cleaning_required':
            return None
        if any(q.get('roomNumber') == room_number for q in self.cleaning_queue):
            return None

        entry = {"roomNumber": room_number, "reason": reason, "addedAt": int(time.time() * 1000)}
        self.cleaning_queue.append(entry)
        self._try_assign_cleaner(entry)
        return entry

    def add_to_cleaning_queue(self, room_number, reason='manual'):
        room = store.get_room(room_number)
        if not room:
            return None
        if room.get('status') == 'occupied':
            logger.warn(f"[HOUSEKEEPING] {room_number}-xona band, navbatga qo'shilmaydi")
            return None
        if room.get('status') in ('cleaning', 'maintenance'):
            return None
        if any(q.get('roomNumber') == room_number for q in self.cleaning_queue):
            return None

        if room.get('status') != 'cleaning_required' and reason == 'guest_checkout':
            logger.debug(f"[HOUSEKEEPING] {room_number} allaqachon {room.get('status')}, navbatga qo'shilmaydi")
            return None

        if room.get('status') != 'cleaning_required':
            old_status = room.get('status')
            store.update_room(room_number, {"status": 'cleaning_required', "dirtyAt": int(time.time() * 1000)})
            broker.publish('room.status_changed', {
                "roomNumber": room_number, "oldStatus": old_status, "newStatus": 'cleaning_required',
                "changedBy": 'housekeeping', "reason": reason,
            })

        entry = {"roomNumber": room_number, "reason": reason, "addedAt": int(time.time() * 1000)}
        self.cleaning_queue.append(entry)
        self._try_assign_cleaner(entry)
        return entry

    def _try_assign_cleaner(self, queue_entry):
        housekeepers = store.get_housekeepers()
        free = next((h for h in housekeepers if h.get('available')), None)
        if not free:
            return None
        queue_entry["assignedTo"] = free['id']
        queue_entry["assignedToName"] = free['name']
        queue_entry["assignedAt"] = int(time.time() * 1000)
        return free

    def start_cleaning(self, room_number, by_user='system'):
        room = store.get_room(room_number)
        if not room:
            return {"success": False, "error": f"{room_number}-xona topilmadi"}
        if room.get('status') != 'cleaning_required':
            return {"success": False, "error": f"{room_number}-xona tozalash uchun tayyor emas (joriy: {room.get('status')})"}
            
        old_status = room.get('status')
        store.update_room(room_number, {
            "status": 'cleaning',
            "cleaningStartedAt": int(time.time() * 1000),
            "cleanedBy": by_user,
        })
        
        broker.publish('room.status_changed', {
            "roomNumber": room_number, "oldStatus": old_status, "newStatus": 'cleaning',
            "changedBy": 'housekeeping', "actor": by_user,
        })
        logger.info(f"[HOUSEKEEPING] {room_number}-xona tozalanmoqda ({by_user})")
        return {"success": True, "room": store.get_room(room_number)}

    def mark_clean(self, room_number, by_user='system'):
        room = store.get_room(room_number)
        if not room:
            return {"success": False, "error": f"{room_number}-xona topilmadi"}
        if room.get('status') not in ('cleaning', 'cleaning_required'):
            return {"success": False, "error": f"{room_number}-xona tozalanmagan (joriy: {room.get('status')})"}
            
        old_status = room.get('status')
        now = int(time.time() * 1000)
        cleaning_duration_ms = now - room.get('cleaningStartedAt', now) if room.get('cleaningStartedAt') else None

        store.update_room(room_number, {
            "status": 'inspection',
            "lastCleanedAt": now,
            "inspectionStartedAt": now,
            "dirtyAt": None,
            "cleaningStartedAt": None,
            "lastCleaningDurationMs": cleaning_duration_ms,
            "lastCleanedBy": by_user,
        })

        self.cleaning_queue = [q for q in self.cleaning_queue if q.get('roomNumber') != room_number]

        store.increment_stat('totalCleanings')

        broker.publish('room.status_changed', {
            "roomNumber": room_number, "oldStatus": old_status, "newStatus": 'inspection',
            "changedBy": 'housekeeping', "actor": by_user,
        })
        broker.publish('notification.created', {
            "type": 'inspection_required',
            "severity": 'info',
            "message": f"{room_number}-xona tozalandi va qabul tomonidan tekshirilishi kutilmoqda",
            "roomNumber": room_number,
        })

        logger.info(f"[HOUSEKEEPING] {room_number}-xona tozalandi -> TEKSHIRUVDA")
        return {"success": True, "room": store.get_room(room_number)}

    def get_queue(self):
        res = []
        for q in self.cleaning_queue:
            item = dict(q)
            item["room"] = store.get_room(q["roomNumber"])
            res.append(item)
        return res

housekeeping_service = HousekeepingService()
